server/etcdserver/server.go

```diff
@@ -980,6 +989,7 @@ func (s *EtcdServer) applyAll(ep *etcdProgress, apply *toApply) {
 	<-apply.notifyc
 
 	s.triggerSnapshot(ep)
+	s.maybeCompactRaftLog(ep)
```

Comment:
If you really want to get this small merged firstly, then please ensure it's as independent as possible.  Currently the `s.snapshot` performs both snapshot and compaction operations. It makes sense to extract the compaction operation as an independent function/method, but let's call the method inside `s.triggerSnapshot`,

```
func (s *EtcdServer) triggerSnapshot(ep *etcdProgress) {
	if !s.shouldSnapshot(ep) {
		return
	}
	lg := s.Logger()
	lg.Info(
		"triggering snapshot",
		zap.String("local-member-id", s.MemberID().String()),
		zap.Uint64("local-member-applied-index", ep.appliedi),
		zap.Uint64("local-member-snapshot-index", ep.snapi),
		zap.Uint64("local-member-snapshot-count", s.Cfg.SnapshotCount),
		zap.Bool("snapshot-forced", s.forceSnapshot),
	)
	s.forceSnapshot = false

	s.snapshot(ep.appliedi, ep.confState)
	ep.snapi = ep.appliedi
	s.compact(xxxxx)   // call the new method here, so we still do it each time after creating a snapshot.

}
```

Reply:
No rush on merging this PR. If we do merge it, we need to ensure etcd actually benefits from it. Let's resolve https://github.com/etcd-io/etcd/issues/17098#issuecomment-2374327464 first.

Reply:
The goal of the PR is to make compaction independent from snapshot. Not just refactor the function.

Reply:
> Not just refactor the function.

Just refactoring the function (extract the compact into a separate method) is an independent and safe change, accordingly can be merged soon.

> The goal of the PR is to make compaction independent from snapshot

It modifies the logic/semantics, so it's no longer an independent change.

Reply:
To be clearer about https://github.com/etcd-io/etcd/pull/18635#discussion_r1775414327, I am proposing an independent & safe minor refactoring below as the very first step

https://github.com/ahrtr/etcd/commit/efae0d29f7ffef1ef743fb11b8936f3f69497cac